File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
// class Solution {
// public:
//     int lengthOfLIS(vector<int>& nums) {
//         int n = nums.size();
//         vector<int> dp(n, 1);
//         for(int i = 0; i < n; i++){
//             for (int j = 0; j < i; j++) {
//                 if (nums[i] > nums[j]) {
//                     dp[i] = max(dp[i], dp[j] + 1);
//                 }
//             }
//         }
        
//         return *max_element(dp.begin(), dp.end());
//     }
// };
class Solution {
private:
    vector<int> dp;
public:
    int solve(int pos, vector<int>& nums) {
        if (pos >= nums.size()) {
            return 1;
        }
        int &ret = dp[pos];
        if (ret != -1) {
            return ret;
        }
        ret = 1;
        for (int i = pos + 1; i < nums.size(); i++) {
            if (nums[pos] < nums[i]) {
                ret = max(ret, 1 + solve(i, nums));
            }
        }
        return ret;
    }
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();     
        dp.resize(n, -1);
        int ans = 0;
        for (int i = 0; i < n;i++) {
            ans = max(ans, solve(i, nums));
        }
        return ans;
    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [-1] * n
        def solve(pos, nums):
            if pos == len(nums):
                return 1
            if dp[pos] != -1:
                return dp[pos]
            res = 1
            for i in range(pos + 1, len(nums)):
                if nums[pos] < nums[i]:
                    res = max(res, 1 + solve(i, nums))
            dp[pos] = res
            return res
        return max([solve(i, nums) for i in range(n)])
    